### ISSUE TITLE
tensorflow and mace produce different output result

### ISSUE THREAD

**Issue by dimitryn** [XiaoMi/mace#238]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.0
- **NDK version(e.g., 15c)**:
17b
- **GCC version(if compiling for host, e.g., 5.4.0)**:
- **MACE version (Use the command: git describe --long --tags)**:
master
- **Python version(2.7)**: 
2.7
- **Bazel version (e.g., 0.13.0)**:
15
### Model deploy file (*.yml)
```yaml
......
```

### Describe the problem
Hi , I have TEST tensorflow model with Conv2D , DepthwiseConv2D, and BiassAdd layers. After converting to mace format it produces BITEXACT output on Google Pixel 2 !!!!!

But when we converted our Real model with additional to test model Relu, FuseBatchNorm, Reshape, Shape, Pack, StridedSlice, Pack, Const layers - results are completely wrong !!!!

Please advise how to resolve it.
Thanks.

P.S.  Latest Mace versoin show very good performance result compares to tensorflow lite on Google Pixel 2 with our model, but this shortstopper preventing form us to move forward.

**Comment by liyinhgqw** [comment]
Can you describe this issue with more details?
Which model did you run on Pixel 2? On CPU or GPU? How do you decide if your results are correct or wrong?

**Comment by dimitryn** [comment]
Hi @liyinhgqw ,
1. Model similar to Resnet50.
2. CPU.
3.Input data same for tensorflow and Mace.

**Comment by liyinhgqw** [comment]
Can you attach your model and yaml config file for us to debug?

**Comment by dimitryn** [comment]
Hi @liyinhgqw ,
download atached files .
Do to size limitation I split zip file in 10 parts .
1. remove zip extension from files Data1.z01.zip - Data1.z10.zip -> Data1.z01 - Data1.z10
2.use winzip to combine all parts  together .
 3.Unzip zip file.
[Data1.z01.zip](https://github.com/XiaoMi/mace/files/2542510/Data1.z01.zip)
[Data1.z02.zip](https://github.com/XiaoMi/mace/files/2542511/Data1.z02.zip)
[Data1.z03.zip](https://github.com/XiaoMi/mace/files/2542512/Data1.z03.zip)
[Data1.z04.zip](https://github.com/XiaoMi/mace/files/2542513/Data1.z04.zip)
[Data1.z05.zip](https://github.com/XiaoMi/mace/files/2542514/Data1.z05.zip)
[Data1.z06.zip](https://github.com/XiaoMi/mace/files/2542515/Data1.z06.zip)
[Data1.z07.zip](https://github.com/XiaoMi/mace/files/2542516/Data1.z07.zip)
[Data1.z08.zip](https://github.com/XiaoMi/mace/files/2542517/Data1.z08.zip)
[Data1.z09.zip](https://github.com/XiaoMi/mace/files/2542518/Data1.z09.zip)
[Data1.z10.zip](https://github.com/XiaoMi/mace/files/2542519/Data1.z10.zip)
[Data1.zip.zip](https://github.com/XiaoMi/mace/files/2542520/Data1.zip.zip)

**Comment by dimitryn** [comment]
Hi @liyinhgqw , any progress ?

**Comment by liyinhgqw** [comment]
Could you resend your models and yaml by email:
liyin@xiaomi.com ?

**Comment by llhe** [comment]
@dimitryn Thanks for report. It's suggested to use Google drive for large files sharing or using free zip tools like 7zip to make splits. Winzip is usually not available for us. (I tried to unzip it using a trail version winzip but failed)

**Comment by dimitryn** [comment]
@llhe I send model and instructions to  @liyinhgqw

**Comment by lee-bin** [comment]
Maybe you can replace following ops with const shape(1,32768)
```text
Resface/Logits/Flatten/flatten/Shape
Resface/Logits/Flatten/flatten/strided_slice
Resface/Logits/Flatten/flatten/Reshape/shape
```
The data format of MACE/CPU/float is NCHW, so we need to transform 
`FormerOp(4D)->Reshape(2D)->MatMul(2D)` to `FC` to convert tensorflow model to MACE model.
Otherwise, input to MatMul would be in wrong format.

**Comment by lee-bin** [comment]
You can try `tf.layers.dense` after Conv4

**Comment by dimitryn** [comment]
@lee-bin  
1.tf.layers.dense  is not an option for us.

2.Replace flatten with const shape as you suggested did not change output, there is still a difference between mace and tensorflow.

3. We try to move all graph into NCHW but there been an error on conversion :
```
Generic conv implementation only supports NHWC tensor format for now:
[[Node: Resface/Conv1/Conv1_pre/Conv2D = Conv2D[T=DT_FLOAT, data_format="NCHW", dilations=[1, 1, 1, 1], padding="SAME", strides=[1, 1, 2, 2], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](transpose, Resface/Conv1/Conv1_pre/weights)]]
```

4. We believe there is a bug in flatten / Matmul  and it does not work.

**Comment by lee-bin** [comment]
1. Did you replace flatten with reshape(batch, height * width * depth)?
MACE/CPU/float converter can transform `FormerOp(4D)->Reshape(2D)->MatMul(2D)` to `FormerOp(4D)->FC(4D)` and perform right, but not `FormerOp(4D)->Flatten(2D)->MatMul(2D)`.
2. `MatMul(2D)->Reshape(2D)->FusedBatchNorm(4D)` is not gonna work either, due to NCHW format.
3. So if you transform to dense and batchnorm, that would work fine on MACE.
4. If that't still not working, you can send us the python code so that we can see if we can help you.

**Comment by dimitryn** [comment]
@lee-bin I attached  flatten replace with reshape
![capture](https://user-images.githubusercontent.com/36672340/48182230-1fbb3580-e333-11e8-9ed8-b8c3b9d10899.PNG)

![screenshot 2](https://user-images.githubusercontent.com/36672340/48182185-f69aa500-e332-11e8-9707-8f7ec833e089.png)

**Comment by dimitryn** [comment]
@lee-bin does shape in the attached image correct?
![screenshot 2](https://user-images.githubusercontent.com/36672340/48182264-395c7d00-e333-11e8-83bb-0a393eeb17b3.png)

**Comment by lee-bin** [comment]
This seems fine.

**Comment by dimitryn** [comment]
@lee-bin Update.
1.After another check MatMul working bitexact with tensorflow , flatten replace with reshape working, Thanks.
2.problem still exist after MatMul  and before final output_features layers.
Any suggestion?

T
![capture](https://user-images.githubusercontent.com/36672340/48185635-55b1e700-e33e-11e8-8200-4894f605e124.PNG)
haks.

**Comment by lee-bin** [comment]
1. As I mentioned, Reshape maybe a problem due to different data formats. So if you can replace `Shape`
after MatMul to const shape and just get FusedBatchNorm's output as result, that would work fine.
2. MACE currently does not remove Reshape after MatMul when transform to FC, so maybe you can send me a new model after replacing Shape with Const shape, so that I can test.
My email is libin11@xiaomi.com.

**Comment by dimitryn** [comment]
thanks for help
